docs/how-tos/create-nexus-files-by-python-and-validate-them.md

```diff
@@ -0,0 +1,940 @@
+
+1. It is located in the Tree at position: ENTRY/definition/
+
+2. The "@" indicates, that this is an attribute of the concept "definition".
+
+3. The name of the attribute is "version".
+
+4. The "required" indicates, that this attribute has to be added to be in line with the NeXus definition "NXoptical\_spectroscopy".
+
+5. The "NX\_CHAR" indicates the datatype. This is should be a string: "The preferred string representation is UTF-8" (more information see here: [https://manual.nexusformat.org/nxdl-types.html](<https://manual.nexusformat.org/nxdl-types.html>))
+
+![image.png](<./attachments/51dc82f9f0f5ec2f-image.png>)
+
+Now the python script has to be extended in the following:
+
+```
+f['/entry/definition'].attrs['version'] = 'v2024.02'
+```
+
+This h5py command extends the entry "/entry/definition" by the attribute named "version" with the value "v2024.02". The same is done for the URL attribute:
+
+```
+f['/entry/definition'].attrs['URL'] = 'https://github.com/FAIRmat-NFDI/nexus_definitions/blob/f75a29836431f35d68df6174e3868a0418523397/contributed_definitions/NXoptical_spectroscopy.nxdl.xml'
+```
+
+Though, you have to use your versions which you want to refer to, as in a few years this NeXus definition might change a little bit. This is shown in the following.
+
+### How to get the "version" and "URL" values
+
+At the time, you create the NeXus definition. Go to the page of the respectively used NeXus concept, i.e. [https://fairmat-nfdi.github.io/nexus\_definitions/classes/contributed\_definitions/NXoptical\_spectroscopy.html](<https://fairmat-nfdi.github.io/nexus_definitions/classes/contributed_definitions/NXoptical_spectroscopy.html>)
+
+Scroll down until you find "**NXDL Source**:" and follow this link, i.e. [https://github.com/FAIRmat-NFDI/nexus\_definitions/blob/fairmat/contributed\_definitions/NXoptical\_spectroscopy.nxdl.xml](<https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXoptical_spectroscopy.nxdl.xml>)
+
+This is the github website, in which the latest NeXus definition of NXoptical\_spectroscopy is stored in the NeXus definition language file (.nxdl). The information is structured in the xml format.
+
+Now you have to copy the permalink of this file. Go to the top right side of the website. Find the Menu made by 3 dots:
+
+![image.png](<./attachments/c6ab2f4b925aed27-image.png>)
+
+Copy the permalink and insert it as value for the "URL" attribute (Step 1, Red box in the image)
+
+Go to "nexus\_definitions" (Step 2, Red box in the image)
+
+![image.png](<./attachments/d8e727b3b32dcbb9-image.png>)
+
+On the right side, you should see below "Releases" the "tags" (Red box in the image). Follow this link.
+
+Copy the latest Tag, which should look similar to "v2024.02". Insert it as value for the "version" attribute.
```

Comment:
yea, exactly. This is better to be technically correct, thouhj I wanted to avoid such technical details as this is not the point of the tutorial.

This may sound funny, but: Clicking these few links on GitHub, will already be confusing enough for most people which only now how to work in a lab (no/few python knowledge).

Reply:
I get that. Still, the created file itself will be not valid because `version="v2024.02"` does not match the nxs appdef that is used here,

Reply:
Hm okay. So, I have to extend it to as well get the latest GitTag manually? This will be too prone to errors.
Not sure how to proceed here.

Reply:
I suggest you leave it as it is, but add a disclaimer that this is not actually correct and one should add the proper version. Then add a link to how it is done in pynxtools (you can even say that his is a reason why people should use our tools).

This would be the link to the function: https://github.com/FAIRmat-NFDI/pynxtools/blob/c13716915bf8f69068c3b94d1423681b580fd437/src/pynxtools/_build_wrapper.py#L17

Reply:
I added:

### Disclaimer:
It would be better, to specify this version tag to include as well the "GitHub commit id". In this way, a [pynxtools generated version tag](https://github.com/FAIRmat-NFDI/pynxtools/blob/c13716915bf8f69068c3b94d1423681b580fd437/src/pynxtools/_build_wrapper.py#L17) might look like this:
`v2022.07.post1.dev1278+g1d7000f4`. For simplicity, this is omitted here.

Hope this suits.

Question:
I as well showed how to copy the permalink, does this in some way as well include the GitHub commit? At least it can be used to uniquely identify the used nexus version?

Reply:
Yeah, sounds good.

No, you will not get the commit id from that link. What you show are _releases_, i.e., these are triggered by NIAC and mark the state of the definitions at a point in time, similar to when we make a new release of our tools, like pynxtools=0.6.0. The `post1.dev1278+g1d7000f4` part that gets added indicates

- .post1: this is a post-release version, meaning an update has been made after the original v2022.07 release. The 1 here suggests this is the first such post-release version.
- .dev1278: this is a development version. The 1278 might indicate this is the 1278th development iteration since the last official release.
- +g1d7000f4:
  - The + symbol typically denotes that this version includes extra metadata.
  - The g is a prefix used to indicate that this is a Git commit identifier.
  - g1d7000f4 is the shortened Git commit hash

This typically uses [PEP440](https://peps.python.org/pep-0440/). Maybe you could add a link to that PEP to show where this comes from.

So, you'll only get that by building the repo from git. I don't think github has the possibility to auto-create these tags.